functions/taskrouterListeners/createHrmContactListener.private.ts

```diff
@@ -0,0 +1,187 @@
+ */
+export const shouldHandle = ({
+  TaskAttributes: taskAttributesString,
+  TaskSid: taskSid,
+  EventType: eventType,
+}: EventFields) => {
+  if (!eventTypes.includes(eventType)) return false;
+
+  const { isContactlessTask, transferTargetType } = JSON.parse(taskAttributesString ?? '{}');
+
+  if (isContactlessTask) {
+    console.debug(`Task ${taskSid} is a contactless task, contact was already created in Flex.`);
+    return false;
+  }
+
+  if (transferTargetType) {
+    console.debug(
+      `Task ${taskSid} was created to receive a ${transferTargetType} transfer. The original contact will be used so a new one will not be created.`,
+    );
+    return false;
+  }
+  return true;
+};
+
+export const handleEvent = async (
+  { getTwilioClient, HRM_STATIC_KEY, TWILIO_WORKSPACE_SID }: Context<EnvVars>,
+  { TaskAttributes: taskAttributesString, TaskSid: taskSid, WorkerSid: workerSid }: EventFields,
+) => {
+  const taskAttributes = taskAttributesString ? JSON.parse(taskAttributesString) : {};
+  const { channelSid } = taskAttributes;
+
+  const client = getTwilioClient();
+  const serviceConfig = await client.flexApi.configuration.get().fetch();
+
+  const {
+    definitionVersion,
+    hrm_base_url: hrmBaseUrl,
+    hrm_api_version: hrmApiVersion,
+    form_definitions_version_url: configFormDefinitionsVersionUrl,
+    assets_bucket_url: assetsBucketUrl,
+    helpline_code: helplineCode,
+    channelType,
+    customChannelType,
+    feature_flags: { enable_backend_hrm_contact_creation: enableBackendHrmContactCreation },
+  } = serviceConfig.attributes;
+  const formDefinitionsVersionUrl =
+    configFormDefinitionsVersionUrl || `${assetsBucketUrl}/form-definitions/${helplineCode}/v1`;
+  if (!enableBackendHrmContactCreation) {
+    console.debug(
+      `enable_backend_hrm_contact_creation is not set, the contact associated with task ${taskSid} will be created from Flex.`,
+    );
+    return;
+  }
+  console.debug('Creating HRM contact for task', taskSid);
+  const hrmBaseAccountUrl = `${hrmBaseUrl}/${hrmApiVersion}/accounts/${serviceConfig.accountSid}`;
+
+  const contactForApi: HrmContact = {
+    ...BLANK_CONTACT,
+    channel: (customChannelType || channelType) as HrmContact['channel'],
+    rawJson: {
+      definitionVersion,
+      ...BLANK_CONTACT.rawJson,
+    },
+    twilioWorkerId: workerSid as HrmContact['twilioWorkerId'],
+    taskId: taskSid as HrmContact['taskId'],
+    channelSid: channelSid ?? '',
+    serviceSid: (channelSid && serviceConfig.chatServiceInstanceSid) ?? '',
+    // We set createdBy to the workerSid because the contact is 'created' by the worker who accepts the task
+    createdBy: workerSid as HrmContact['createdBy'],
+  };
+
+  const prepopulatePath = Runtime.getFunctions()['hrm/prepopulateForm'].path;
+  const { prepopulateForm } = require(prepopulatePath) as PrepopulateForm;
+  await prepopulateForm(taskAttributes, contactForApi, formDefinitionsVersionUrl);
```

Comment:
I think is harder to follow what is really happening here, cause `prepopulateForm` is mutating the given contact. Could we return a new object to make this explicit? (I will really appreciate that information being evident next time I'm around this code :stuck_out_tongue:)

Reply:
I won't make it a pure function because that makes it significantly uglier for deep objects and there's no practical benefit here since it's a brand new contact anyway.

I thought using 'populate' in the method name would have made the mutation effect clear, bad assumption on my part.

So the function still mutates but we reassign the output to a new variable to be clear. I also changed the method name to be more descriptive

Reply:
The reassignment is good enough, I don't mind it being pure as much as being explicit that the contact was transformed. Thanks for the change!